pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+	logger.Debugf("putting object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+	_, err = s3client.PutObjectInBucket(bucket, string(s3HealthCheckObjectBody), s3HealthCheckObjectKey, contentType)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to put object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Get the object from the bucket
+	logger.Debugf("getting object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+	read, err := s3client.GetObjectInBucket(bucket, s3HealthCheckObjectKey)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Compare the old and the existing object
+	logger.Debug("comparing objects md5")
+	oldMD5 := md5.Sum([]byte(s3HealthCheckObjectBody))
+	currentMD5 := md5.Sum([]byte(read))
+	if currentMD5 != oldMD5 {
+		err = errors.Wrapf(err, "wrong file content, old file MD5 is %x and new file MD5 is %x", oldMD5, currentMD5)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	logger.Debug("successfully checked object store endpoint")
+
+	// Update the EndpointStatus in the CR to reflect the healthyness
+	updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionConnected, "")
+
+	return nil
+}
+
+func cleanupObjectHealthCheck(s3client *S3Client, objectStoreUID string) error {
+	bucketToDelete := genUniqueBucketName(objectStoreUID)
+	logger.Infof("deleting object %q from bucket %q", s3HealthCheckObjectKey, bucketToDelete)
+	_, err := s3client.DeleteObjectInBucket(bucketToDelete, s3HealthCheckObjectKey)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *rgwStatusChecker) cleanupHealthCheck() error {
+	bucketToDelete := genUniqueBucketName(c.objContext.UID)
+	logger.Infof("deleting object %q from bucket %q", s3HealthCheckObjectKey, bucketToDelete)
+
+	_, err := DeleteBucket(c.objContext, bucketToDelete, true)
+	if err != nil {
+		logger.Errorf("failed to delete bucket %q. %v", bucketToDelete, err)
+	}
+
+	userToDelete := c.genUserConfig()
+	output, err := DeleteUser(c.objContext, userToDelete.UserID)
+	if err != nil {
+		logger.Errorf("failed to delete s3 user %q. %s. %v", userToDelete.UserID, output, err)
+	} else {
+		logger.Debugf("successfully deleted s3 user %q", userToDelete.UserID)
+	}
+
+	return nil
+}
+
+func toCustomResourceStatus(currentStatus *cephv1.EndpointStatus, details string) *cephv1.EndpointStatus {
+	s := &cephv1.EndpointStatus{
+		LastChecked: time.Now().UTC().Format(time.RFC3339),
+		Details:     details,
+	}
+
+	if currentStatus != nil {
+		s.LastChanged = currentStatus.LastChanged
+		if currentStatus.Details != s.Details {
+			s.LastChanged = s.LastChecked
+		}
+	}
+	return s
+}
+
+func genUniqueBucketName(uuid string) string {
```

Comment:
Ah here is the uniqueness per cluster. How about unique object instead of unique bucket?

Reply:
I believe we should have a dedicated bucket/user per cluster since the permissions for a given user are set to a particular bucket. Having the same bucket/user for all the cluster is likely to cause issues. Especially on the external mode when the CR gets deleted, the bucket/user are expected to go away and I don't want to leave anything behind.